dex/src/main/scala/com/wavesplatform/dex/AddressActor.scala

```diff
@@ -287,43 +288,52 @@ class AddressActor(owner: Address,
       addressWsMutableState = addressWsMutableState.removeSubscription(client)
       context.unwatch(client)
 
+    // Received a snapshot for pending connections
     case SpendableBalancesActor.Reply.GetSnapshot(allAssetsSpendableBalance) =>
       allAssetsSpendableBalance match {
         case Right(spendableBalance) =>
           addressWsMutableState.sendSnapshot(
             balances = mkWsBalances(spendableBalance),
             orders = activeOrders.values.map(WsOrder.fromDomain(_))(collection.breakOut),
           )
-          if (!addressWsMutableState.hasActiveSubscriptions) scheduleNextDiffSending
+          if (!addressWsMutableState.hasActiveSubscriptions) scheduleNextDiffSending()
           addressWsMutableState = addressWsMutableState.flushPendingSubscriptions()
         case Left(matcherError) =>
           addressWsMutableState.pendingSubscription.foreach { _.unsafeUpcast[WsServerMessage] ! WsError.from(matcherError, time.correctedTime()) }
           addressWsMutableState = addressWsMutableState.copy(pendingSubscription = Set.empty)
       }
 
+    // It is time to send updates to clients. This block of code asks balances
     case WsCommand.PrepareDiffForWsSubscribers =>
       if (addressWsMutableState.hasActiveSubscriptions) {
         if (addressWsMutableState.hasChanges) {
           spendableBalancesActor ! SpendableBalancesActor.Query.GetState(owner, addressWsMutableState.getAllChangedAssets)
-        } else scheduleNextDiffSending
+          // We asked balances for current changedAssets and clean it here,
+          // because there are could be new changes between sent Query.GetState and received Reply.GetState.
+          addressWsMutableState = addressWsMutableState.cleanBalanceChanges()
+        } else scheduleNextDiffSending()
       }
 
+    // It is time to send updates to clients. This block of code sends balances
     case SpendableBalancesActor.Reply.GetState(spendableBalances) =>
-      if (addressWsMutableState.hasActiveSubscriptions) {
-        addressWsMutableState = addressWsMutableState.sendDiffs(
+      addressWsMutableState = if (addressWsMutableState.hasActiveSubscriptions) {
+        scheduleNextDiffSending()
+        addressWsMutableState.sendDiffs(
           balances = mkWsBalances(spendableBalances),
           orders = addressWsMutableState.getAllOrderChanges
         )
-        scheduleNextDiffSending
-      }
+      } else if (addressWsMutableState.pendingSubscription.isEmpty) {
+        addressWsMutableState.cleanBalanceChanges() // There are neither active, nor pending connections
```

Comment:
Neither active, nor pending - why only balance changes clean? Why not all?

Reply:
Actually all changes are cleaned: orders are cleaned below